proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+>string will be a copy of one of the arguments. 
+
+The import adapter for `chooser_` has to lift the two argument `string`s and
+lower the return value:
+
+```wasm
+(@interface func $stralloc-i
+  (param $str string)
+  (result i32)           ;; return a shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-i"     ;; create string entity of right size
+  let (local $ptr i32) (result i32)
+    local.get $str
+    local.get $ptr
+	local.get $str
+	string.size
+    utf8.from.string
+	local.get $ptr
+  end
+)
+
+(@interface implement (import "" "chooser_")
+  (param $l i32)
+  (param $r i32)
+  (result i32)
+
+  local.get $l
+  call-export "access-utf8-i" ;; get at the text & length of the left string
+  string.from.utf8
+
+  local.get $r
+  call-export "access-utf8-i" ;; the right string
+  string.from.utf8
+
+  call-import "chooser"  ;; leaves a string on stack
+
+  invoke-func "stralloc-i" ;; allocate a local string for the result
+)
+```
+
+Compared to the export adapter, the import adapter is very straightforward. This
+is because we require the caller -- a core wasm function -- to take
+responsibility for the argument strings and for the returned string.
+
+## Fusing adapters
+
+The fused adapter consists of the inlined export adapter within the import
+adapter; which is then simplified. In this case we have a simple import adapter
+being combined with one that has some complexity:
+
+```wasm
+(@interface implement (import "" "chooser_")
+  (param $l i32)
+  (param $r i32)
+  (result i32)
+
+  local.get $l
+  call-export "access-utf8-i"
+  string.from.utf8
+
+  local.get $r
+  call-export "access-utf8-i"
+  string.from.utf8
+
+  let ($left string
+       $right string) (result string)
+    local.get $left
+    invoke-func $stralloc-x
+    invoke-func $shared-x     ;; make it a shared ptr
+    let (local $lp owned<i32>) (result string)
+      local.get $right
+      invoke-func $stralloc-x
+	  invoke-func $shared-x   ;; make it a shared ptr
+      let (local $rp owned<i32>) (result string)
+        ;; set up the call to the core chooser 
+        local.get $lp
+		owned.access
+        local.get $rp
+		owned.access
+        call-export "nondeterministic_choice"  ;; call the chooser itself
+        own (i32)               ;; own the result
+          call-export "shared-release-x" 
+        end
+        owned.access
+        call-export "shared-get-x"
+        call-export "access-utf8-x"
+        string.from.utf8
+      end
+	end	
+  end
+  invoke-func "stralloc-i" ;; allocate a local string for the result
+)
+```
+
+After we expand the helper functions, we get:
+
+```wasm2
+(@interface implement (import "" "chooser_")
+  (param $l i32)
+  (param $r i32)
+  (result i32)
+
+  local.get $l
+  call-export "access-utf8-i"
+  string.from.utf8
+
+  local.get $r
+  call-export "access-utf8-i"
+  string.from.utf8
+
+  let ($left string
+       $right string)
+    local.get $left
```

Comment:
It'd be nice to bracket these inlined functions somehow, maybe just with `;; invoke-func $str-alloc-x`, etc.

Reply:
I have taken out a lot of the intermediate stages. Will write this process up separately.  The final code is not always immediately recognizable anymore - so labeling is problematic.